src/passes/CastRefining.cpp

```diff
@@ -0,0 +1,284 @@
+          // is relevant, if nothing is ever created of the others or their
+          // subtypes.
+          for (auto subType : typeSubTypes) {
+            if (createdTypesOrSubTypes.count(subType)) {
+              if (!refinedType) {
+                // This is the first relevant thing, and hopefully will remain
+                // the only one.
+                refinedType = subType;
+              } else {
+                // We've seen more than one as relevant, so we have failed to
+                // find a singleton.
+                refinedType = std::nullopt;
+                break;
+              }
+            }
+          }
+        }
+        if (refinedType) {
+          // Propagate anything from the child, to handle chains.
+          auto iter = refinableTypes.find(*refinedType);
+          if (iter != refinableTypes.end()) {
+            *refinedType = iter->second;
+          }
+
+          refinableTypes[type] = *refinedType;
+        }
+      }
+    }
+
+    // We found optimizable types. Apply them.
+    Optimizer optimizer(*this);
+    optimizer.run(getPassRunner(), module);
+
+    // Refinalize, as RefCasts may have new types now.
+    ReFinalize().run(getPassRunner(), module);
+  }
+
+  struct Optimizer : public WalkerPass<PostWalker<Optimizer>> {
+    bool isFunctionParallel() override { return true; }
+
+    CastRefining& parent;
+
+    Optimizer(CastRefining& parent) : parent(parent) {}
+
+    std::unique_ptr<Pass> create() override {
+      return std::make_unique<Optimizer>(parent);
+    }
+
+    std::optional<HeapType> getHeapTypeIfRelevant(Type type) {
+      if (type == Type::unreachable || !type.isRef()) {
+        return std::nullopt;
+      }
+
+      // We ignore basic types, as in any non-trivial module they will always
+      // have something created of their subtypes. TODO Optimize even there
+      auto heapType = type.getHeapType();
+      if (heapType.isBasic()) {
+        return std::nullopt;
+      }
+
+      return heapType;
+    }
+
+    template<typename T> void refineCast(T* curr) {
+      auto& castType = curr->getCastType();
+      if (auto heapType = getHeapTypeIfRelevant(castType)) {
+        auto iter = parent.refinableTypes.find(*heapType);
+        if (iter != parent.refinableTypes.end()) {
+          // We can refine this cast.
+          castType = Type(iter->second, castType.getNullability());
+        }
+      }
+    }
+
+    void visitRefCast(RefCast* curr) {
+      auto castType = curr->getCastType();
+      if (auto heapType = getHeapTypeIfRelevant(castType)) {
+        if (parent.createdTypesOrSubTypes.count(*heapType) == 0) {
+          // Nothing is created of this type or any subtype, so the cast can
+          // only pass through a null, at most.
+          Builder builder(*getModule());
+          Expression* rep = nullptr;
+          if (castType.isNullable()) {
+            // TODO: Without TNH we can still optimize here, to do a null check
```

Comment:
This could be as simple as this:
```
curr->castType() = Type(heapType->getBottom(), Nullable);
```
Maybe worth doing in the same PR?

Reply:
Good point, we can just do that here. And we can do it in TNH mode too, and leave the other part for OptimizeInstructions (I'll verify we already do it, or add it if not).